### PR TITLE
Issue #8 - Extended algorithm for sound-playing tabs

### PR DESCRIPTION
**Pull request by ThomasGreiner**
Unfortunately, all these small changes are necessary for the audio algorithm to work properly. About two thirds of the code in this PR is related to tests though.
- Implemented audio-based algorithm
  - Implemented attention gathering for background tabs
  - Added handling for relevant events
  - Determine thresholds based on URLs
  - Updated storage
    - Introduced "audible" and "muted" properties on tab pages in memory
    - Introduced "isAudio" property on tab pages in memory and pages in the database
    - Introduced timers for audio-playing tabs
    - Moved `tabPages.set()` logic into `storage.updatePage()` to centralize updates of tab pages
- Removed existing video thresholds
  - Don't determine thresholds based on entity
  - Removed list of video sites
- Don't store "scrolled-ongoing" and "audible-ongoing" events in database
  _These events can be inferred from "scrolled-start" and "scrolled-end" or "audible" and "muted" respectively. So far storing them wasn't an issue since it required user interaction to trigger "scrolled-ongoing" but the newly introduced "audible-ongoing" will be triggered automatically and can therefore unnecessarily bloat the data we record._
- Updated tests
  - Introduced `prepareExpectObject()` to simplify object comparison
  - Added support for passed arguments for `setInterval()` and `setTimeout()` to session runner
  - Added tests for new functionality

### CODE REVIEW

File: src/lib/background/session/storage.js [hunk 1]
```diff
@@ -148,25 +148,60 @@ exports.removePage = removePage;
...
+    let entity = getEntity(url);
+
+    // Keep tab page notification around as long as domain doesn't change
+    if (tabPage.notification && entity != tabPage.entity)
+    {
+      tabPage.notification = null;
+    }
+
+    tabPage.attention = 0;
+    tabPage.entity = getEntity(url);
```
Comment (erikvold) [inline src/lib/background/session/storage.js]: nit `tabPage.entity = entity`

File: src/lib/background/session/storage.js [hunk 2]
```diff
@@ -148,25 +148,60 @@ exports.removePage = removePage;
...
+  tabPage = tabPages.get(tabId) || tabPage;
+
+  if (!("url" in tabUpdate))
   {
-    ({url} = oldTabPage);
+    ({url} = tabPage);
   }
 
   if (!url)
     return removePage(tabId);
```
Comment (erikvold) [inline src/lib/background/session/storage.js]: seems like we can rewrite this to be one conditional statement.

    if ("url" in tabUpdate)
    {
      // ... 
    }
    else
    {
      ({url} = tabPage);
      if (!url)
        return removePage(tabId);
    }

File: src/lib/background/session/storage.js [hunk 3]
```diff
@@ -148,25 +148,60 @@ exports.removePage = removePage;
  * Update one or more tab page properties
  * @param {number} tabId
  * @param {Object} tabUpdate
- * @param {string} tabUpdate.title
- * @param {string} tabUpdate.url
+ * @param {string} [tabUpdate.title]
+ * @param {string} [tabUpdate.url]
```
Comment (erikvold) [inline src/lib/background/session/storage.js]: looks like we're adding a `isAudio` option to `tabUpdate` here, so we should add a comment about it.

Comment (ThomasGreiner) [reply src/lib/background/session/storage.js]: Done. You're right, I missed that.

**Comment by erikvold** [comment]
LGTM

**Pull request merged**